crates/sovereign-sdk/module-system/sov-modules-stf-blueprint/src/lib.rs

```diff
@@ -574,26 +604,165 @@ where
             }
         }
     }
+
+    fn apply_soft_confirmations_from_sequencer_commitments(
+        &self,
+        sequencer_public_key: &[u8],
+        initial_state_root: &Self::StateRoot,
+        pre_state: Self::PreState,
+        mut da_data: Vec<<Da as DaSpec>::BlobTransaction>,
+        mut witnesses: std::collections::VecDeque<Vec<Self::Witness>>,
+        mut slot_headers: std::collections::VecDeque<Vec<<Da as DaSpec>::BlockHeader>>,
+        validity_condition: &<Da as DaSpec>::ValidityCondition,
+        mut soft_confirmations: std::collections::VecDeque<Vec<SignedSoftConfirmationBatch>>,
+    ) -> (
+        Self::StateRoot,
+        Vec<u8>, // state diff
+    ) {
+        // First extract all sequencer commitments
+        // Ignore broken DaData and zk proofs. Also ignore ForcedTransaction's (will be implemented in the future).
+        let mut sequencer_commitments: Vec<SequencerCommitment> = vec![];
+        for blob in da_data.iter_mut() {
+            // TODO: get sequencer da pub key
+            if blob.sender().as_ref() == sequencer_public_key {
+                let da_data = DaData::try_from_slice(blob.verified_data());
+
+                if let Ok(DaData::SequencerCommitment(commitment)) = da_data {
+                    sequencer_commitments.push(commitment);
+                }
+            }
+        }
+
+        // Then verify these soft confirmations.
+
+        let mut current_state_root = initial_state_root.clone();
+
+        for sequencer_commitment in sequencer_commitments.iter() {
+            // should panic if number of sequencer commitments and soft confirmations don't match
```

Comment:
https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.zip_eq

Reply:
Or check length above +  .zip from std

Reply:
didn't like zip for multiple iter's sytax